lib/puppet/node/facts.rb

```diff
@@ -116,8 +116,25 @@ def add_timestamp
     @timestamp = Time.now
   end
 
+  def to_yaml
+    facts_to_display = Psych.parse_stream(YAML.dump(self))
+    quote_special_strings(facts_to_display)
+  end
+
   private
 
+  def quote_special_strings(fact_hash)
+    fact_hash.grep(Psych::Nodes::Scalar).each do |node|
+      next unless node.value =~ /:/
+
+      node.plain  = false
+      node.quoted = true
+      node.style  = Psych::Nodes::Scalar::DOUBLE_QUOTED
+    end
+
+    fact_hash.yaml
+  end
```

Comment:
There is unfortunately some tech debt around our YAML generation. There are places where we write a YAML hash to a file using `Puppet::Util::Yaml.dump`. For example the `transactionstorefile` contains current and desired values of resources so if puppet remediates an issue we can know if the manifest changed or the system drifted. If the current or desired value for a resource was ever a macaddress, we'd have the same problem that this PR is fixing. Here are all of the places we currently call `Puppet::Util::Yaml.dump`:

```
lib/puppet/indirector/yaml.rb:      Puppet::Util::Yaml.dump(request.instance, file)
lib/puppet/transaction/persistence.rb:    Puppet::Util::Yaml.dump(@new_data, Puppet[:transactionstorefile])
lib/puppet/util/storage.rb:      Puppet::Util::Yaml.dump(@@state, Puppet[:statefile]))
```

I think ideally, we'd add a [`Puppet::Network::FormatSupport#to_yaml` similar to `to_json`, etc](https://github.com/puppetlabs/puppet/blob/main/lib/puppet/network/format_support.rb#L95-L106) and have it call:

```ruby
def to_yaml(*args)
  Puppet::Util::Yaml.dump(to_data_hash, *args)
end
```

And then add the sanitization logic in your PR to `Puppet::Util::Yaml`. That said I'm fine doing that as a separate issue, as I think the facter macaddress is something we need to fix immediately.